### DISCUSSION TITLE
Can we use storybook to test UI for developing open-sourced  react UI components ?

### DISCUSSION THREAD

**Discussion by ManacyJohn** [storybookjs/storybook#25753] [Help]
### Summary

As I reviewed the documentation, I came across the statement:
'storybook init is not made for empty projects
Storybook needs to be installed into a project that is already set up with a framework'

I'm curious if there is any recommended approach to use storybook for testing open source react/vue UI components ?

**Comment by Marklb** [comment]
Could you provide more explanation of what you are asking?

Are you asking if you can use Storybook to test components that you are developing or someone else developed and released as open source? Your title says developing components, which makes me think you are developing the components, but specifying that they are open source and pointing out how the docs say Storybook isn't for empty projects then makes me think you are wanting to test components from an open-source project that isn't yours.

My assumption is that you will have a project that uses open source components as a dependency. So you would not necessarily be testing the dependency component, but if there is a problem with the dependency then your component's test will fail, which still sort of tests the dependency. I could be wrong about my assumption, so more information may help clarify what you are asking.

**Comment by ManacyJohn** [comment]
I am developing a component that will be published as an open source ui component. I started with an empty repo and I'm adding the relevant packages as needed. In order to test the UI, I usually use npm link or an alternative at the end of the development process. I wanted to see if it's possible to use storybook to see how the UI of the components looks before using it in some other project. I wanted to know the steps to follow when adding storybook and setting it up, when not using any starter templates like cra or vite, like in my case.

**Reply by Marklb** [reply]
I think it would be easiest to create a project that you initialize Storybook in, even if it isn't the original project. Storybook doesn't build the components as a completely independent task. The manager, that contains the Storybook UI and addon panels, is fairly independent, but the preview build is sort of a mix of the Storybook code and your project code. Someone more familiar with the builders could probably explain it better, but to make sure your project's components are built correctly Storybook tries to use your project config/build scripts to configure it's builder.

If you initialize Storybook without a project then it doesn't know if it is going to be using React/Angular/Vue/WebComponent/etc components, which may also have specific differences within the same framework as well. By the time you worked out all the needed dependencies/config, necessary for Storybook to build your component, you would probably have a mostly configured project anyway.

Is there a reason you don't want to add Storybook to your project, instead of trying to use it externally? I find it very useful as a development/documentation/testing tool. Most of my components have a `*.stories.ts` file in the same folder that is useful to get an idea how that component is intended to be used and if I want to see it rendered then I just start Storybook and look at those stories. By including Storybook in your project, you are not affecting the normal build of your project, because you just exclude the `*.stories.(ts|js)` files, just like you would exclude `*.(test|spect).(ts|js)` files.

If you do create Storybook in an empty project and use `npm link` to add your project as a dependency, you will miss out on a lot of the documentation and automatic configuration of controls/actions, since most of the information needed for that is typically removed from the built files that get published.
